instrumentation/runtime-telemetry/runtime-telemetry-java8/javaagent/src/main/java/io/opentelemetry/instrumentation/javaagent/runtimemetrics/java8/JarAnalyzer.java

```diff
@@ -0,0 +1,224 @@
+
+  @GuardedBy("this")
+  private Integer jarsPerSecond;
+
+  @GuardedBy("this")
+  private Worker worker;
+
+  private JarAnalyzer() {}
+
+  /** Get the {@link JarAnalyzer} singleton. */
+  public static JarAnalyzer getInstance() {
+    return INSTANCE;
+  }
+
+  /**
+   * Configure the {@link JarAnalyzer}. If not called, {@link #handle(ProtectionDomain)} is a noop.
+   *
+   * @throws IllegalStateException if called multiple times
+   */
+  public synchronized void configure(int jarsPerSecond) {
+    if (this.jarsPerSecond != null) {
+      throw new IllegalStateException("JarAnalyzer has already been configured");
+    }
+    this.jarsPerSecond = jarsPerSecond;
+    this.handler = this::handleInternal;
+  }
+
+  /**
+   * Install {@link OpenTelemetry} and start processing jars if {@link #configure(int)} was called.
+   *
+   * @throws IllegalStateException if called multiple times
+   */
+  public synchronized void maybeInstall(OpenTelemetry unused) {
+    if (worker != null) {
+      throw new IllegalStateException("JarAnalyzer has already been installed");
+    }
+    if (this.jarsPerSecond == null) {
+      return;
+    }
+    // TODO(jack-berg): Use OpenTelemetry to obtain EventEmitter when event API is stable
+    EventEmitter eventEmitter =
+        GlobalEventEmitterProvider.get()
+            .eventEmitterBuilder(JmxRuntimeMetricsUtil.getInstrumentationName())
+            .setInstrumentationVersion(JmxRuntimeMetricsUtil.getInstrumentationVersion())
+            .setEventDomain(EVENT_DOMAIN_PACKAGE)
+            .build();
+    this.worker = new Worker(eventEmitter, toProcess, jarsPerSecond);
+    Thread workerThread =
+        new DaemonThreadFactory(JarAnalyzer.class.getSimpleName() + "_WorkerThread")
+            .newThread(worker);
+    workerThread.start();
+  }
+
+  /**
+   * Identify the JAR associated with the {@code protectionDomain} and queue it to be processed if
+   * its the first time we've seen it.
+   *
+   * <p>NOTE: does nothing if {@link #configure(int)} has not been called.
+   */
+  void handle(ProtectionDomain protectionDomain) {
+    this.handler.accept(protectionDomain);
+  }
+
+  private void handleInternal(ProtectionDomain protectionDomain) {
+    if (protectionDomain == null) {
+      return;
+    }
+    CodeSource codeSource = protectionDomain.getCodeSource();
+    if (codeSource == null) {
+      return;
+    }
+    URL jarUrl = codeSource.getLocation();
+    if (jarUrl == null) {
+      return;
+    }
+    URI locationUri;
+    try {
+      locationUri = jarUrl.toURI();
+    } catch (URISyntaxException e) {
+      logger.log(Level.WARNING, "Unable to get URI for jar URL: " + jarUrl, e);
+      return;
+    }
+
+    if (!seenUris.add(locationUri)) {
+      return;
+    }
+    if ("jrt".equals(jarUrl.getProtocol())) {
+      logger.log(Level.FINEST, "Skipping processing jar for java runtime module: " + jarUrl);
+      return;
+    }
+    if (!jarUrl.getFile().endsWith(JarAnalyzerUtil.JAR_EXTENSION)) {
```

Comment:
Have you tried to figure out whether there are non-jar urls that could be relevant? I didn't verify but I'd guess that osgi bundles might not point to jar here, war files will probably just point to `WEB-INF/classes` directory on most app servers etc. Could enable this temporarily (or only for smoke tests) and see what urls won't get processed for smoke tests.

Reply:
I think its likely that there are other types of locations that warrant trying to handle. Treating logging as a discovery mechanism: 
- if we find other situations which are not useful, we can add logic to intentionally skip them.
- if we find other situations and deem them useful, can add logic to handle them and any challenges.

Reply:
To enable this for smoke tests add it to https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/bebd81927771e60954582854ec494f117fd8b600/smoke-tests/src/test/java/io/opentelemetry/smoketest/AbstractTestContainerManager.java#L20-L44

Reply:
I enabled this for smoke tests and used the output to refine the processing logic a bit:
- There were code locations that I'm calling "non-archive" based. These are locations that end in `/classes/` or `/jsp/`. I don't know of anyway we can extract / send metadata for these directories, so skipping them with a `FINEST` log message.
- There are `.war` files that I was ignoring. Adjusted the logic to include these.

Here's some filtered output from the smoke tests that show the attributes we manage to extract: https://gist.github.com/jack-berg/d20898c73aa6269d7fd1f766006db90a